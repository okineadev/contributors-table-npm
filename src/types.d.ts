import type { Endpoints } from '@octokit/types'

export type Contributor = Partial<
	Omit<
		Endpoints['GET /repos/{owner}/{repo}/contributors']['response']['data'][number],
		'avatar_url'
	> & {
		login: string
		avatar_url: string
	}
>
