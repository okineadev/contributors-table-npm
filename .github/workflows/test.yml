name: ğŸ§ª Tests

on:
  # https://blog.esciencecenter.nl/reduce-reuse-recycle-save-the-planet-one-github-action-at-a-time-4ab602255c3f
  push:
    paths:
      - 'src/**'
      - 'tests/**'
      - 'package.json'
      - 'bun.lock'
      - 'tsconfig.json'

  pull_request:
    paths:
      - 'src/**'
      - 'tests/**'
      - 'package.json'
      - 'bun.lock'
      - 'tsconfig.json'

  workflow_dispatch:

permissions:
  contents: read

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - name: ğŸ“¥ Checkout
        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
        with:
          persist-credentials: false

      - name: ğŸ”§ Setup Bun
        uses: oven-sh/setup-bun@735343b667d3e6f658f44d0eca948eb6282f2b76 # v2
        with:
          bun-version: latest

      - name: ğŸ“¦ Install dependencies
        run: bun install --frozen-lockfile

      - name: ğŸ§ª Run tests
        run: bun run test
