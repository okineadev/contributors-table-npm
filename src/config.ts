import type { Config } from 'svgo'

/**
 * Application authentication that is presented
 * with each request to the GitHub API
 */
// TODO: Find out if it is necessary
export const USER_AGENT =
	'contributors-table-generator (https://github.com/okineadev/contributors-table-npm)'

/** SVGO configuration for SVG compression */
export const SVGOConfig: Config = {
	plugins: [
		{
			name: 'preset-default',
			params: {
				overrides: {
					removeUnknownsAndDefaults: false,
					removeTitle: false,
				},
			},
		},
	],
}
