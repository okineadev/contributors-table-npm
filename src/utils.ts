import sharp from 'sharp'
import { USER_AGENT } from './config'

/**
 * Fetches an image from the given URL and converts it to a base64-encoded string.
 *
 * @param url - The URL of the image to fetch.
 * @returns A promise that resolves to a base64-encoded string representation of the image.
 */
export async function getbase64Image(url: string) {
	const response = await fetch(url, {
		headers: {
			'User-Agent': USER_AGENT,
		},
	})
	const contentType = response.headers.get('content-type') || 'image/png'
	const buffer = await response.arrayBuffer()
	return `data:${contentType};base64,${Buffer.from(buffer).toString('base64')}`
}

/**
 * Converts an SVG string to a PNG buffer.
 *
 * @param SVG - The SVG string to be converted.
 * @returns A promise that resolves to a buffer containing the PNG image.
 */
export async function generatePNGFromSVG(SVG: string) {
	const svgBuffer = Buffer.from(SVG)
	const pngBuffer = await sharp(svgBuffer).png().toBuffer()

	return pngBuffer
}
