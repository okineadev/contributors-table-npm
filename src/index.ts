import { optimize } from 'svgo'

import { SVGOConfig } from './config'
import { getbase64Image, generatePNGFromSVG } from './utils'
import type { Contributor } from './types'

interface ContributorsTableParams {
	/** Horizontal spacing between avatars in pixels */
	gap?: number
	/** Width and height of each avatar in pixels */
	width?: number
	/** Number of avatars per row */
	columns?: number
	/** Number of rows */
	rows?: number
	/** Border radius of avatars in pixels or `'yes'` for full roundness (width value) */
	roundness?: number | 'yes'
	/** Width of the border around avatars in pixels */
	strokeWidth?: number
	/** Whether to use server-side rendering to fetch and embed avatars in the SVG */
	ssr?: boolean
	/** Output image type*/
	format?: 'svg' | 'png'

	/** @deprecated use `format` instead */
	type?: 'svg' | 'png'
	/** @deprecated use `strokeWidth` instead */
	borderWidth?: number
}

/**
 * Generates image containing table with contributor avatars with links to their GitHub profiles
 *
 * @param contributors - Array of contributor objects containing `login` and `avatar_url`
 * @param options - Configuration options
 *
 * @returns Promise resolving to SVG string or PNG buffer
 */
export async function generateContributorsTable<
	T extends ContributorsTableParams,
	ReturnFormat = T['format'] extends 'png' ? Buffer : string,
>(
	contributors: Contributor[],
	{
		gap = 6,
		width = 40,
		columns = 21,
		rows = 7,
		roundness = 6,
		strokeWidth = 0,
		ssr = true,
		format = 'svg',

		type = undefined,
		borderWidth = undefined,
	}: T = {} as T,
): Promise<ReturnFormat> {
	if (contributors.length === 0) {
		throw new Error('The list of contributors is empty')
	}

	strokeWidth = borderWidth ?? strokeWidth
	format = type ?? format

	const adjustedRoundness =
		typeof roundness === 'string' && roundness === 'yes' ? width : roundness

	rows = rows || Math.ceil(contributors.length / columns)
	// const actualColumns = Math.min(columns, params.contributors.length)

	const svgDimensions = {
		width: columns * width + strokeWidth + (columns - 1) * gap,
		height: rows * width + strokeWidth + (rows - 1) * gap,
	}

	let SVG = `<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="${svgDimensions.width}" height="${svgDimensions.height}">`

	const svgStyle = `
	<style>
		a {
			cursor: pointer;
		}
		a > svg {
			overflow: visible;
		}
		a > svg > rect {
			stroke: #c0c0c0;
			stroke-width: ${strokeWidth ? `${strokeWidth}px` : 0};
			width: ${width}px;
			height: ${width}px;
		}
	</style>
	`

	SVG += svgStyle

	// Sequential avatar loading
	const avatarPromises = []
	for (const contributor of contributors) {
		const avatarUrl =
			ssr || format === 'png'
				? await getbase64Image(`${contributor.avatar_url}&s=${width}`)
				: `${contributor.avatar_url}&amp;s=${width}`
		avatarPromises.push(avatarUrl)
	}

	const avatarUrls = await Promise.all(avatarPromises)

	contributors.forEach((contributor, index) => {
		const username = contributor.login
		const avatarUrl = avatarUrls[index]

		const avatarPosition = {
			x: (index % columns) * (width + gap) + strokeWidth / 2,
			y: Math.floor(index / columns) * (width + gap) + strokeWidth / 2,
		}

		SVG += `
		<a href="https://github.com/${username}">
		  <svg x="${avatarPosition.x}" y="${avatarPosition.y}">
			<title>${username}</title>
			<rect rx="${adjustedRoundness}" fill="url(#i${index})"></rect>

			<defs>
			  <pattern id="i${index}" patternUnits="userSpaceOnUse" width="${width}" height="${width}">
				<image href="${avatarUrl}" width="${width}" height="${width}"></image>
			  </pattern>
			</defs>
		  </svg>
		</a>`
	})

	SVG += '</svg>'

	if (format === 'png') {
		return generatePNGFromSVG(SVG) as ReturnFormat
	}

	return optimize(SVG, SVGOConfig).data as ReturnFormat
}
