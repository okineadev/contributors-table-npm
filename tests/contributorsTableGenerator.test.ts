import { describe, it, expect } from 'bun:test'
import { generateContributorsTable } from '../src'
import sampleContributor from '../samples/sampleContributor'

// Only one contributor
const contributors = [sampleContributor]

const sampleParams = {
	ssr: false,
	width: 40,
}

describe('generateContributorsTable', () => {
	it('should generate SVG with correct dimensions and attributes', async () => {
		const result = await generateContributorsTable(
			Array(2).fill(contributors[0]),
			sampleParams,
		)
		expect(result).toMatchSnapshot()
	})

	it('should generate PNG buffer when type is png', async () => {
		const result = await generateContributorsTable(contributors, {
			...sampleParams,
			format: 'png',
		})
		expect(result).toBeInstanceOf(Buffer)
	})

	it('should render the correct number of contributors', async () => {
		const result = (await generateContributorsTable(
			Array(2).fill(contributors[0]),
			sampleParams,
		)) as string
		expect(result.match(/<title>contributor<\/title>/g)?.length).toBe(2)
	})

	it('should throw an error when the list of contributors is empty', async () => {
		expect(generateContributorsTable([], sampleParams)).rejects.toThrow(Error)
	})

	it('should render avatars on server-side when ssr is true', async () => {
		const result = await generateContributorsTable(contributors, {
			...sampleParams,
			ssr: true,
		})
		expect(result).toContain('data:image/png;base64,')
	})

	it('should apply full roundness when roundness is "yes"', async () => {
		const result = await generateContributorsTable(contributors, {
			...sampleParams,
			roundness: 'yes',
		})
		expect(result).toContain(`rx="${sampleParams.width}"`)
	})
})
