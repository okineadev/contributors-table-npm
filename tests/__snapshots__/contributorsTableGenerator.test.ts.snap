// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`generateContributorsTable should generate SVG with correct dimensions and attributes 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="960" height="40"><a href="https://github.com/contributor" style="cursor:pointer"><svg x="0" y="0" style="overflow:visible"><title>contributor</title><rect fill="url(#a)" rx="6" style="stroke:silver;stroke-width:0;width:40px;height:40px"/><defs><pattern id="a" width="40" height="40" patternUnits="userSpaceOnUse"><image width="40" height="40" href="https://avatars.githubusercontent.com/u/583231?v=4&amp;s=40"/></pattern></defs></svg></a></svg>"`;
